### ISSUE TITLE
Not working with hot module replacement.

### ISSUE THREAD

**Issue by theKashey** [alvarotrigo/react-fullpage#18]
Any changes, caused HMR, are breaking something inside full page, and it's stop working.

Example - https://codesandbox.io/s/m5qjp6q0o9

- First - everything is working
- Then edit anything, and codesandbox will hot-update example
- fullpage is partially dead.

**Comment by cmswalker** [comment]
Thanks for reporting this, have you noticed this behavior outside of codesandbox?

**Comment by theKashey** [comment]
Yep, using fullpage with React-Hot-Loader causes 2 slides per-scroll after any update.
Ie FP is working, but skipping odd slides.

**Comment by alvarotrigo** [comment]
> No, this is something absolutely different.

Mmm it might be we are having exactly the same issue. I noticed fullpage.js scrolls twice on your codepen when updating the title of the first section, as you said. 
However, if I open the developer tools in Chrome after that, I can see the following:
![screen shot 2018-08-23 at 12 25 58](https://user-images.githubusercontent.com/1706326/44520330-df530080-a6cf-11e8-9512-505a7f642542.png)

But, if I open the JS console before updating the title of the first section, I can see [the error I talked about previously](https://github.com/alvarotrigo/react-fullpage/issues/18#issuecomment-414610347).

So apparently, the same memory leak error but different outputs depending on whether the console is open or closed ?

Also, sometimes I won't get any error on the console, but then fullpage.js will be kind of frozen. And scrolling up or down won't change sections. 

@cmswalker apparently the error might be related with the callback. In [line 125](https://github.com/alvarotrigo/react-fullpage/blob/master/components/ReactFullpage/index.js#L125):

```javascript
        case 'onLeave':
          state = makeState(fromArgs(['index', 'nextIndex', 'direction']));
          break;```

This is what I get when trying to debug:
![screen shot 2018-08-23 at 12 44 02](https://user-images.githubusercontent.com/1706326/44521749-6efaae00-a6d4-11e8-89a3-5f9b5e57fc36.png)

**Comment by alvarotrigo** [comment]
I updated my last reply. The issue is related with `this.setState(state);`

**Comment by theKashey** [comment]
So - in short - behavior is strange.

**Comment by yourmediafriend** [comment]
I have a similar issue to @theKashey using redux and react-router. I am not using hot module.  I have several fullpage layouts. The first one works fine but the next one skips 2 slides at a time. I believe the origin is not reset when mounting new instance of component. 

also I am not sure if it is related but if I go from a fullpage layout to a page of scrollable text the body scroll is disabled.  On investigation it would appear the html class and body styling add by the module are not removed.   

I have only recently started with react-router but maybe withRouter might help .

**Comment by cmswalker** [comment]
@theKashey regarding react-router it sounds like you're using multiple instances of fullpage. That seems fine to me so long as the components are unmounting properly, but I'm not sure if `fullpage.js` supports more than 1 instance of `fullpage` at a time, is this currently supported @alvarotrigo couldn't find any issues/docs relating to it in the readme of `fullpage.js`

My first guess is that when the router is unmounting one fullpage for another, the internal state of the original one persists into the 2nd one, thus why the slide skipping occurs (it's trying to catch up and is off)

**Comment by yourmediafriend** [comment]
@cmswalker does the module have a destroy / reset function that handles unmounting ? this would be useful especially in the case of body and html tag classes.

**Comment by theKashey** [comment]
👨‍💻where is `componentWillUnmount`? 
It's actually a case for codeSandbox or router - __one__ full page at single point of time(hot module replacement) - redraw all components without page reloading -  the old FP "still" active, as was long not properly destroyed.

**Comment by cmswalker** [comment]
Looks like we'll need to add a call to `componentWillUnmount` but first we'll need to make a teardown/reset function in https://github.com/alvarotrigo/fullPage.js in order to destroy all listeners so we can rebuild again. Basically, we'll need both instance based fullpage support (which is accomplished by tearing down old instances and their handlers) @alvarotrigo does this sound doable?

**Comment by alvarotrigo** [comment]
> I'm not sure if fullpage.js supports more than 1 instance of fullpage at a time, is this currently supported 

Nop. Fullpage.js only supports a single instance.

**Comment by alvarotrigo** [comment]
> @alvarotrigo couldn't find any issues/docs relating to it in the readme of fullpage.js

Good point! Although you should get the following error on the JS console when trying it:

>  Fullpage.js can only be initialized once and you are doing it multiple times!

**Comment by cmswalker** [comment]
hmmm, seems to fully support react lifecycle methods we'd need multiple instances. Basically when calling `fullpage_api.destroy('all')` that would put us into a state where a new instance could be generated. Otherwise we won't be able to support those using fullpage on multiple routes of a single page application (unless we hand the instance back to them and they can re register it to another dom node but that seems odd/inconvenient)

How much work would it be to allow fullpage.js to support instances? It doesn't necessarily need to support multiple instances at once, but it would need to support 1 brand new instance taking the place of another (kinda like a singleton that is either alive or destroyed, so to speak vs 100 instances) Does this make sense?

**Comment by alvarotrigo** [comment]
> Basically when calling `fullpage_api.destroy('all')` that would put us into a state where a new instance could be generated. 

@cmswalker yeah, and that's the case.
When using `fullpage_api.destroy('all')` you should be able to initialise fullpage.js again. The previous "instance" gets destroyed.

Is that what you need?

**Comment by cmswalker** [comment]
Ok great, I'm thinking https://github.com/alvarotrigo/react-fullpage/pull/25 will allow support for routing, and will hopefully help with some HMR issues

**Comment by cmswalker** [comment]
@yourmediafriend could you try out v0.0.9 and see if this fixes your routing issues.

**Comment by cmswalker** [comment]
@theKashey can you reproduce the codesandbox issue on v0.0.9?

**Comment by theKashey** [comment]
👍 Working for me 👍

**Comment by yourmediafriend** [comment]
All good. Appreciated.
